### ISSUE TITLE
Pass localization not working after fields push

### ISSUE THREAD

**Issue by timmy4all** [alexandercerutti/passkit-generator#23]
Hi,
I am setting the pass fields dynamically from code and then trying to localize the labels. I do not have lproj directory created as the media is still the same.
My pass.json has passtype as
`generic : {}`

I am setting values as 
```
examplePass.primaryFields.push({
        "key" : "coupID",
        "label" : "CoupId_localized",
        "value" : "testid"
      });
```

Then localize as follows for french
```
examplePass.localize("fr", {
      "CoupId_localized": "Copon num_FR"
    });

examplePass.localize("en", {
      "CoupId_localized": "Copon num"
    });
```

The file gets dowloaded but its corrupted.

**Comment by alexandercerutti** [comment]
Hi @timmy4all, thank you for the report!
I'm sure this is a bug, but I'll investigate better this evening and update you 😉

**Comment by alexandercerutti** [comment]
I'm investigating right now, @timmy4all. I tried to reproduce what you said and got all working fine.
May I ask you to provide me the generated pass? If you prefer you can also send me via email or Telegram, at https://t.me/AlexandrCerutti

**Comment by timmy4all** [comment]
Thanks @alexandercerutti . I can't email you the pass due to confidentiality but below is the code and pass.json i am using. To be clear, the pass downloads succesfully without errors. When i email the pass to my ios device and try to open attachment it prompts to select which app to open. If i remove the localization it automaticallly shows the pass when you tap the attachment and allows to add to wallet.

`var createError = require('http-errors');
var express = require('express');
var path = require('path');
var cookieParser = require('cookie-parser');
var logger = require('morgan');

var indexRouter = require('./routes/index');
var usersRouter = require('./routes/users');
const { Pass } = require('passkit-generator');

var app = express();
// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'pug');
app.use(logger('dev'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));
app.use('/', indexRouter);
app.use('/users', usersRouter);

  app.get('/downloadpass', async (req, res, next) => {
    let examplePass = new Pass({
      model: "./passModels/testCard",
        certificates: {
          wwdr: "./keys/wwdr.pem",
          signerCert: "./keys/signerCert.pem",
          signerKey: {
            keyFile: "./keys/signerKey.pem",
            passphrase: "TEST123456"
          }
        }
    });
    examplePass.primaryFields.push(
      {
        "key" : "TestID",
        "label" : "TestID",
        "value" : "999111999"
      }
    );
    try {
     let stream = await examplePass.generate();
     res.set({
			"Content-type": "application/vnd.apple.pkpass",
			"Content-disposition": `attachment; filename=myPass.pkpass`
		});
     stream.pipe(res)
   } catch (e) {
     //this will eventually be handled by your error handling middleware
     console.log(e) 
   }
   })

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});`

And pass.json.
`{
  "formatVersion": 1,
  "passTypeIdentifier" : "pass.com.sample.dev.testcard",
  "teamIdentifier" : "QJ62W1SB8Z",
  "serialNumber": "SKR03KD",
  "relevantDate": "2011-12-08T13:00-08:00",
  "locations": [
    {
      "longitude": -122.3748889,
      "latitude": 37.6189722
    },
    {
      "longitude": -122.03118,
      "latitude": 37.33182
    }
  ],
  "organizationName" : "Test ORG",
  "description" : "Test ORG Card",
  "logoText" : "Test Organization",
  "foregroundColor" : "rgb(0, 0, 0)",
  "backgroundColor" : "rgb(243,226,226)",
  "labelColor": "rgb(0, 0, 0)",
  "generic": {}
}`

Can you share a sample project with dynamic fields, pass model and localization working? I can try to reproduce with your code.

**Comment by alexandercerutti** [comment]
Okay, no problem. I understand that passes may contain confidential infos but I request passes only for debugging purposes and delete them once the issue is closed.

I took one of the examples ([`barcode.js`](https://github.com/alexandercerutti/passkit-generator/blob/master/examples/barcode.js)), and replaced the code from Line 31 to Line 72 with the code you passed me (primary fields and localization) and changed `examplePass` to `Pass`.

I also edited examples model `examplePass` to not have `generic: {}` as you told me.

Yeah, the download usually happen correctly even if the mime-type is wrong. A test you can try to do is to open up in a Mac or PC and rename it to `.zip`.
If it opens correctly, the pass is well composed but might have other internal problems (some violated rule for example).

Also, if you have a Mac (thing that I don't have right now), you can connect your iPhone via USB and open `Console.app` (/Applications/Utility/Console.app). You'll be able to see iPhone logs, also the ones when you try to open a non-valid Pass. You can filter your log search for "Passkit".

I don't see anything wrong with your pass.json

I've edited your comment to look better at the code. 😉

**Comment by timmy4all** [comment]
@I ran the test by connecting my device to mac. 
The console error when i tried to open the pass is as follows
```
error	09:39:24.762629 -0400	MobileMail	Invalid data error reading pass pass.com.sample.dev.testcard/SKR03KD. No entry in manifest for file en.lproj/pass.strings
error	09:39:24.762675 -0400	MobileMail	Invalid data error reading pass pass.com.sample.dev.testcard/SKR03KD. No entry in manifest for file en.lproj/pass.strings

```

**Comment by alexandercerutti** [comment]
The only weird thing I see is the double slash (\\) in the paths, but I don't know if it is normal and accepted or not.
Sadly I cannot say to try to edit the manifest manually because signature is based on it.

The model I tested, as I said before, is available in `examples/models/examplePass.pass` here on Github;
I edited it to reflect you fields (as I told you, I left only `generic: {}` instead of `eventTicket: { ... }`).

**Comment by timmy4all** [comment]
Could you share the compiled pass? I can compare the files to find the difference. Isn't the framework generating the manifest though?

**Comment by alexandercerutti** [comment]
I'll share it this evening: I don't have it right now 😉
Yes, it is. But I never noticed those two slashes.

**Comment by timmy4all** [comment]
I generated the example pass using localization.js from your examples and got the same error.

`error	10:55:23.849043 -0400	MobileMail	Invalid data error reading pass pass.com.sample.dev.testcard/SKR03KD. No entry in manifest for file de.lproj/pass.strings
`

**Comment by alexandercerutti** [comment]
I really don't know what's going on....
Can I ask you again for the pass you generated? Maybe I can find more infos...
Different people already passed me theirs for the support and I deleted them after the issue, so you have nothing to worry about. I never share them. I promise.

I'll send you mine this evening anyway, so we can test each other the results. What do you think?

**Comment by timmy4all** [comment]
I will send you the one i generated with your example. Can you share your email?

**Comment by alexandercerutti** [comment]
If you have telegram, you can send it to me to @AlexandrCerutti. Otherwise <surname>.<name>@gmail.com 😉

**Comment by alexandercerutti** [comment]
Okay, @timmy4all. Initially I wasn't able to open it because I had key `transitType` key, which is not valid in passes not of type "boardingPass".
Then I had obliviously the mismatch between type identifier, team identifier and certificate ones.
But then it went all fine...
I cannot really understand what's going on.

Can you try to start your process with "DEBUG=*" first? `DEBUG=* node app.js` and tell me if it says anything about the library when you try to generate the pass? It will print also a lot of thing from express. Thank you.

**Comment by timmy4all** [comment]
Thanks,

Can you send me your compiled pass that  is working with  localization? I
want to try opening it on my phone and see if it works.

**Comment by alexandercerutti** [comment]
Sure, I'll send it to you back via email.

**Comment by timmy4all** [comment]
Hey alexander,
Thanks for sending me the pass. Yes you are right, the pass you generated does not have double slash in the manifest.
I tried by hard coding the primary field in pass json and simply using localize for fr and en like you did and i still have the double slash in the manifest.
Can you send me the code you used, maybe il try exact code base to generate same pass that you did.

**Comment by alexandercerutti** [comment]
Sure. I'll just send the file. It requires `express` installed. I suggest you to install it globally.
Also, the file `test.js` should be placed in examples. I attached also webserver.js and the model that works for me. If you place `test.js` in the `examples` folder, you won't need `webserver.js`

[test.zip](https://github.com/alexandercerutti/passkit-generator/files/3204584/test.zip)

Let me know.

**Comment by timmy4all** [comment]
Hey alexander. 
I think i understand a bit how to reproduce this problem.
I tried my same code and model on Mac and it worked fine.
If i run in on my windows laptop it adds // instead of / in the manifest.
I wonder if its the "path" library in the pass class of the framework thats the culprit here.

**Comment by alexandercerutti** [comment]
@timmy4all Ehe, I was thinking that too...
Probably I don't get any problem because I use WSL to develop under Windows...
I have to find a better way to join them, if I can find one. Maybe forcing "/" 🤔🤔

**Comment by timmy4all** [comment]
I have tried printing the path strings from pass.js during generation and it does not have the double slash.
Its happening sometime after 
`const stringFilePath = path.join(`${l}.lproj`, "pass.strings");` - here i still see single / when print to console.

**Comment by alexandercerutti** [comment]
Mhhh, good catch! Could it be something related to `JSON.Stringify()` maybe? I'm supposing...

**Comment by alexandercerutti** [comment]
Okay, I discovered a very interesting thing. I tried to execute the example I passed you through node.js on Windows and found that "\\" character is automatically escaped when it is used in a key name. I don't know exactly the reason, because if I try to reproduce it both in node and browser consoles, it gets stripped if one, or two if you try to escape it.

Probably because it is used as an escape sequence (like `\n` for example). Therefore there might be a "security system" (I don't know which one) to automatically use it twice.

The problem here is always related to the `path.join` used before, because we shouldn't use it to join it with "\" but with "/".
